File: python/bowling/bowling.py
class BowlingGame:
    def __init__(self):
        self.frames = []
        self.cur_frame = (None, None)

    def is_strike(self, frame):
        return frame[0] == 10

    def is_spare(self, frame):
        return not self.is_strike(frame) and sum(frame) == 10

    def roll(self, pins):
        f = self.frames
        cur = self.cur_frame
        if pins > 10 or pins < 0 or (cur[0] and pins + cur[0] > 10): raise ValueError("too high")
        if len(f) == 11 and not self.is_strike(f[-1]) and not self.is_spare(f[-1]): raise ValueError("invalid")
        if len(f) == 10 and not self.is_strike(f[-1]) and not self.is_spare(f[-1]): raise ValueError("invalid")
        if len(f) == 10 and self.is_spare(f[-1]) and cur[0] is not None: raise ValueError("invalid")
        if len(f) == 11 and not self.is_strike(f[-2]) and not self.is_spare(f[-1]): raise ValueError("invalid")
        if pins == 10:
            f.append((10, 0))
        elif cur[0] is not None:
            if cur[0] + pins > 10: raise ValueError("invalid")
            f.append((cur[0], pins))
            self.cur_frame = (None, None)
        else:
            self.cur_frame = (pins, None)


    def score(self):
        f = self.frames
        if self.cur_frame[0] is not None: f.append((self.cur_frame[0], 0))
        if len(f) < 10: raise ValueError("invalid")
        total = 0
        for i, frame in enumerate(f):
            if i >= 10: break
            total += sum(frame)
            if frame[0] == 10:
               _first = f[i + 1][0]
               _second = f[i + 2][0] if _first == 10 else f[i + 1][1]
               total += _first + _second
            elif sum(frame) == 10:
               total += f[i + 1][0]

        return total

File: ruby/bowling/bowling.rb
class Game
  LAST_FRAME = 9

  def initialize
    @open_frame = nil
    @bonus = Array.new(0)
    @frames = []
  end

  def score
    raise BowlingError unless game_complete?
    (0..LAST_FRAME).map { |i|
      frame_score(i)
    }.sum
  end

  def game_complete?
    return false if @frames.size < 10

    last_is_strike = strike?(@frames[LAST_FRAME])
    last_is_spare  = spare?( @frames[LAST_FRAME])
    return true unless last_is_strike or last_is_spare
    return true if last_is_strike and @bonus.size == 2
    return true if last_is_spare  and @bonus.size == 1

    return false
  end

  def frame_score(i)
    raise BowlingError unless game_complete?

    frame = @frames[i]
    score = frame.sum
    if i == LAST_FRAME
      score += @bonus.sum
    elsif strike?(frame)
      next_frame = @frames[i+1]
      score += next_frame.sum
      score += (i == 8 ? @bonus[0] : @frames[i+2][0]) if strike?(next_frame)
    elsif spare?(frame)
      score += @frames[i+1][0]
    end
    score
  end

  def roll(pins)
    raise BowlingError unless pins.between?(0,10)
    raise BowlingError if game_complete?

    if @open_frame
      raise BowlingError if @open_frame + pins > 10
      @frames << [@open_frame, pins]
      @open_frame = nil
    elsif @frames.size == 10
      b1 = @bonus[0]
      raise BowlingError if b1 and b1 != 10 and b1 + pins > 10
      @bonus << pins
    elsif pins == 10
      @frames << [pins, 0]
    else
      @open_frame = pins
    end
  end

  def strike?(frame)
    frame == [10, 0]
  end

  def spare?(frame)
    !strike?(frame) && frame.sum == 10
  end

  class BowlingError < StandardError; end
end
